### PR TITLE
[spirv] Add a pipeline to use workgroup memory

### PR DESCRIPTION
**Pull request by antiagainst**
This commit adds a pipeline for using workgroup memory in
SPIR-V CodeGen. It basically moves existing LLVM GPU
code into Common/ directory and use them. This pipeline
is meant to support desktop/server class GPUs where using
shared memory can give nice performance gains. So along
the way, added configuration stubs for NVIDIA/AMD GPUs. 
They are not tuned yet, just to get started.

### CODE REVIEW

**Review by MaheshRavishankar — COMMENTED** [review]
Rather move the `GPUUtils` to `Utils` folder.

File: iree/compiler/Codegen/Passes.h [hunk 1]
```diff
@@ -391,6 +392,8 @@ void addSPIRVTileAndVectorizePassPipeline(OpPassManager &pm);
 /// performs distribution to threads with vectorization.
 void addSPIRVTileAndVectorizeToCooperativeOpsPassPipeline(OpPassManager &pm);
 
+void addSPIRVTileAndVectorizeWithWorkgroupMemoryPassPipeline(OpPassManager &pm);
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Passes.h:395]: Please add comments.

File: iree/compiler/Codegen/Passes.td [hunk 2]
```diff
@@ -163,6 +163,17 @@ def MemrefCopyToLinalgPass :
       "mlir::iree_compiler::createMemrefCopyToLinalgPass()";
 }
 
+def GPUDistributeSharedMemoryCopy :
+    Pass<"iree-codegen-gpu-distribute-shared-memory-copy", "func::FuncOp"> {
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Passes.td:167]: Make this `iree-gpu-*` (and the one below too)

File: iree/compiler/Codegen/SPIRV/SPIRVTile.cpp [hunk 3]
```diff
@@ -31,6 +31,8 @@
 namespace mlir {
 namespace iree_compiler {
 
+static constexpr int kNumGPUDims = 3;
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/SPIRV/SPIRVTile.cpp:34]: Is this even used anywhere?

File: iree/compiler/Codegen/SPIRV/Utils.cpp [hunk 4]
```diff
@@ -22,6 +22,8 @@
 namespace mlir {
 namespace iree_compiler {
 
+static constexpr int kNumGPUDims = 3;
```
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/SPIRV/Utils.cpp:25]: Is this used anywhere?

**Review by MaheshRavishankar — COMMENTED** [review]
Skimmed through mostly and minor comments here. 

Overall I was hoping this could be the impetus to move the promotion logic, etc. to use tensor -> vectors. This is going to reach EOL pretty soon.... But happy yo turn the functionality on and then pivot if that is easier.

Looks fine to me, but will Thomas approve.

File: iree/compiler/Codegen/SPIRV/Utils.cpp (hunk 4 above)
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/Utils.cpp]: Yes used later. Now changed to getting the definition from GPUUtils.h.

File: iree/compiler/Codegen/SPIRV/SPIRVTile.cpp (hunk 3 above)
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/SPIRVTile.cpp]: Yes used later. Now changed to getting the definition from GPUUtils.h.

File: iree/compiler/Codegen/Passes.h (hunk 1 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Passes.h:399]: Done.

File: iree/compiler/Codegen/SPIRV/KernelConfig.cpp [hunk 5]
```diff
@@ -537,7 +553,7 @@ static LogicalResult setSPIRVOpConfig(const spirv::TargetEnv &targetEnv,
         // Try to tile and vectorize first.
         std::array<int64_t, 2> workgroupXY = {32, 2};
         std::array<int64_t, 3> threadMNK = {8, 8, 4};
-        auto result = detail::setMatmulOpConfig(op, workgroupXY, threadMNK);
+        auto result = detail::setMatmulOpConfig(op, 32, workgroupXY, threadMNK);
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/SPIRV/KernelConfig.cpp:556]: nit: maybe move 32 in a variable with a descriptive name or add a comment?

File: iree/compiler/Codegen/Utils/GPUUtils.cpp [hunk 6]
```diff
@@ -65,5 +67,52 @@ std::array<int64_t, 3> getWorkgroupSize(mlir::func::FuncOp funcOp) {
   return workgroupSize;
 }
 
+bool canPerformVectorAccessUsingAllThreads(ArrayRef<int64_t> shape,
```
Comment (ThomasRaoux) [inline iree/compiler/Codegen/Utils/GPUUtils.cpp:70]: note that those functions are super ad hoc and are likely to change and break in the future. It is fine to re-use but just want to point it out if you had different expectations.

**Review by ThomasRaoux — APPROVED** [review]

File: iree/compiler/Codegen/Utils/GPUUtils.cpp (hunk 6 above)
Comment (antiagainst) [reply iree/compiler/Codegen/Utils/GPUUtils.cpp:70]: SG, thanks for the reminder!

File: iree/compiler/Codegen/SPIRV/KernelConfig.cpp (hunk 5 above)
Comment (antiagainst) [reply iree/compiler/Codegen/SPIRV/KernelConfig.cpp]: done

**Pull request merged**